source/io/module_eggShell.js

```diff
@@ -572,20 +571,30 @@
             return EggShell_GetArrayDimLength(Module.eggShell.v_userShell, vi, path, dim);
         };
 
-        Module.eggShell.resizeArray = publicAPI.eggShell.resizeArray = function (vi, path, newDimensionSizes) {
-            var int32Byte = 4;
-            var rank = newDimensionSizes.length;
-            var newLengths = Module._malloc(rank * int32Byte);
-
-            for (var i = 0; i < rank; i += 1) {
-                Module.setValue(newLengths + (i * int32Byte), newDimensionSizes[i], 'i32');
+        Module.eggShell.resizeArray = publicAPI.eggShell.resizeArray = function (valueRef, newDimensions) {
+            if (!Array.isArray(newDimensions)) {
+                throw new Error('Expected newDimensions to be an array of dimension lengths, instead got: ' + newDimensions);
             }
-
-            var success = EggShell_ResizeArray(Module.eggShell.v_userShell, vi, path, rank, newLengths);
-
-            Module._free(newLengths);
-
-            return success;
+            var stack = Module.stackSave();
+            var newDimensionsLength = newDimensions.length;
+            var dimensionsPointer = Module.stackAlloc(newDimensionsLength * LENGTH_SIZE);
+            var i, currentDimension;
+            for (i = 0; i < newDimensionsLength; i += 1) {
+                currentDimension = newDimensions[i];
+
+                if (typeof currentDimension !== 'number') {
+                    throw new Error('Expected all dimensions of newDimensions to be numeric values for dimension length, instead got' + currentDimension);
+                }
+                Module.setValue(dimensionsPointer + (i * LENGTH_SIZE), currentDimension, 'i32');
+            }
+            var eggShellResult = Module._EggShell_ResizeArray(Module.eggShell.v_userShell, valueRef.typeRef, valueRef.dataRef, newDimensionsLength, dimensionsPointer);
+            if (eggShellResult !== 0) {
+                throw new Error('Resizing the array failed for the following reason: ' + eggShellResultEnum[eggShellResult] +
+                ' (error code: ' + eggShellResult + ')' +
+                ' (typeRef: ' + valueRef.typeRef + ')' +
+                ' (dataRef: ' + valueRef.dataRef + ')');
+            }
+            Module.stackRestore(stack);
```

Comment:
Do you know if any of the Errors thrown above would cause a memory leak?

Reply:
Written under the assumption that on exception throw to JS the whole Vireo instance (stack + heap) should be considered corrupt.

We should verify that assumption is usable and standardize across functions prior to merge. Tracked as "Robustness of stackSave / restore and error recovery the api should enable" in #466